silence/silence.go

```diff
@@ -0,0 +1,792 @@
+// Copyright 2016 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The silence package provides a storage for silences, which can share its
+// state over a mesh network and snapshot it.
+package silence
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/matttproud/golang_protobuf_extensions/pbutil"
+	pb "github.com/prometheus/alertmanager/silence/silencepb"
+	"github.com/prometheus/common/model"
+	"github.com/satori/go.uuid"
+	"github.com/weaveworks/mesh"
+)
+
+// ErrNotFound is returned if a silence was not found.
+var ErrNotFound = fmt.Errorf("not found")
+
+func utcNow() time.Time {
+	return time.Now().UTC()
+}
+
+// Silences holds a silence state that can be modified, queried, and snapshottet.
+type Silences struct {
+	logger    log.Logger
+	now       func() time.Time
+	retention time.Duration
+
+	gossip mesh.Gossip // gossip channel for sharing silences
+
+	// We store silences in a map of IDs for now. Currently, the memory
+	// state is equivalent to the mesh.GossipData representation.
+	// In the future we'll want support for efficient queries by time
+	// range and affected labels.
+	mtx sync.RWMutex
+	st  gossipData
+}
+
+// Options exposes configuration options for creating a new Silences object.
```

Comment:
Yes, will add that. That seems to say everything as all the defaults are no-ops.